Add MOON baseline

QinbinLi opened the pull request:
<!--
Thank you for opening a pull request (PR)!

Contribution guidelines: https://github.com/adap/flower/blob/main/CONTRIBUTING.md
-->

## Issue

### Description

<!--
Describe the problem addressed by this PR.

Example: The variable name `rnd` could be misinterpreted as an abbreviation of *random*, but it refers to the current server round.
-->

### Related issues/PRs

<!--
Link issues and/or PRs that are related to this PR.

Example: Fixes #123. See also #456 and #789.
-->

## Proposal

### Explanation

<!--
Explain the changes and how they improve the issue described above.

Example: The variable `rnd` was renamed to `server_round` to improve readability.
-->

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### Any other comments?

<!--
Please be aware that it may take some time until the maintainers can review the PR.
Smaller PRs with good descriptions can be considered much more easily.

If you have an urgent request or question, please use the Flower Slack:

    https://flower.dev/join-slack/ (channel: #contributions)

Thank you for contributing to Flower!
-->

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,129 @@
+---
+title: Model-Contrastive Federated Learning
+url: https://arxiv.org/abs/2103.16257
+labels: [data heterogeneity, image classification]
```
```suggestion
labels: [data heterogeneity, image classification, cross-silo, constrastive-learning]
```

jafermarq replied:
does it make sense adding those extra labels?

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,129 @@
+> Note: If you use this baseline in your work, please remember to cite the original authors of the paper as well as the Flower paper.
+
+
+****Paper:**** :https://arxiv.org/abs/2103.16257
```
```suggestion
****Paper:**** : [arxiv.org/abs/2103.16257](https://arxiv.org/abs/2103.16257)

```

jafermarq commented on baselines/moon/README.md:
````diff
@@ -0,0 +1,129 @@
+
+```bash
+# install the base Poetry environment
+poetry install
+
+# activate the environment
+poetry shell
+
+# install PyTorch with GPU support.
+pip install torch==1.12.0+cu116 torchvision==0.13.0+cu116 torchaudio==0.12.0 --extra-index-url https://download.pytorch.org/whl/cu116
````
Let's add a bit more detail here:
```
# set local python version via pyenv
pyenv local 3.10.6
# then fix that for poetry
poetry env use 3.10.6
# then install poetry env
poetry install
```

As you'll see on a comment for `pyproject.toml`, I moved the PyTorch installation there. (so everything needed is installed with just `poetry install` 🪄 )

jafermarq commented on baselines/moon/README.md:
````diff
@@ -0,0 +1,129 @@
+
+First ensure you have activated your Poetry environment (execute `poetry shell` from this directory). To run MOON on CIFAR-10 (Table 1 of the paper), you should run:
+```bash  
+python -m moon.main cifar10 
````
```suggestion
python -m moon.main --config-name cifar10 
```

jafermarq replied:
I think all these commands need to use the `--config-name` flag to tell Hydra to look for a specific `.yaml` config other than that set in `main.py` as the default (i.e. `base.yaml`)

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,129 @@
+
+****Contributors:**** : Qinbin Li
+
+## Experimental Setup
```
Could you add at the bottom of this section a short paragraph describing briefly how the "stateful" clients are implemented? and why it is needed for MOON? (if you think it's worth pointing to a specific part of the paper, then that's great too)

jafermarq commented on baselines/moon/moon/conf/cifar10.yaml:
```diff
@@ -0,0 +1,33 @@
+  # model config
+  name: simple-cnn
+  output_dim: 256
+  dir: ./models/moon/cifar10/
```
```suggestion
  dir: ./client_states/moon/cifar10/
```

jafermarq replied:
Maybe we can rename this directory? (if you want to go ahead, please apply this change to the other configs)

jafermarq commented on baselines/moon/moon/client.py:
```diff
@@ -0,0 +1,159 @@
+    ) -> Tuple[NDArrays, int, Dict]:
+        """Implement distributed fit function for a given client."""
+        self.set_parameters(parameters)
+        if self.prev_net is None:
+            self.prev_net = init_net(self.dataset, self.model, self.output_dim)
+            self.prev_net = copy.deepcopy(self.net)
+        else:
+            # load previous model from model_dir
+            self.prev_net.load_state_dict(
+                torch.load(
+                    os.path.join(self.model_dir, str(self.net_id), "prev_net.pt")
+                )
```
Flower clients in simulation are ephemeral. This means that all their internal variables etc are discarded once the client completes the tasks it was spawned for (in this case to do `fit()`). This means that all the time, the above `else` is never happening (since `self.prev_net` is always None). I'd advice to change the if/else and instead load the model if a file is found with the matching name. Does it make sense?

Built-in support for stateful clients in Flower is almost ready. So later (after the _Summer of Reproducibility_ we'll update this logic)

jafermarq commented on baselines/moon/moon/conf/cifar10.yaml:
```diff
@@ -0,0 +1,33 @@
+
+client_resources:
+  num_cpus: 4
+  num_gpus: 1
```
VRAM utilisation is fairly low (~900MB on my machine) for the CIFAR-10 model. I'm wondering if you could leave a lower value by default (e.g. `num_gpus=0.2`) to at least allow for 5 clients to run concurrently in the same GPU. More info on how to set `num_gpus` and `num_cpus` can be found in our docs: https://flower.dev/docs/framework/how-to-monitor-simulation.html. Adjusting these `client_resources` does make a big impact on how fast your experiments run.

jafermarq commented on baselines/moon/moon/conf/cifar100_100clients.yaml:
```diff
@@ -0,0 +1,33 @@
+
+client_resources:
+  num_cpus: 8
+  num_gpus: 1
```
Unlike for CIFAR-10, the `ResNet50` model here needs more VRAM (I'm seeing ~5GB). How about if we set the default to `num_gpus=0.5` instead? (so 2 flower clients share a GPU)

jafermarq reviewed (commented):
Hi @QinbinLi,

I did a first pass through your `MOON` implementation using Flower. I was able to run the code, so that's great! I have left a few comments here and there. Mostly small things. 

More critical is the way you are loading `prev_model` in the client's `fit()` method. It's not doing what I think you want to do (i.e. use the model the client sent back to the server in the previous round). Please take a look at my suggestion below on how to fix it.

Less critical, but also important:
* Would it be possible to generate two plots showing `fedprox` vs `moon`?  (as in Figures 6a,b in your paper)
* Please rename the directory where the figures are added to `_static` (this is helpful to make your README to appear automatically in the documentation).

jafermarq commented on baselines/moon/pyproject.toml:
```diff
@@ -0,0 +1,141 @@
+hydra-core = "1.3.2" # don't change this
+scikit-learn = "1.3.0"
+matplotlib = "3.8.0"
+
```
```suggestion
torch = { url = "https://download.pytorch.org/whl/cu116/torch-1.12.0%2Bcu116-cp310-cp310-linux_x86_64.whl"}
torchvision = { url = "https://download.pytorch.org/whl/cu116/torchvision-0.13.0%2Bcu116-cp310-cp310-linux_x86_64.whl"}
```

jafermarq commented on baselines/moon/pyproject.toml:
```diff
@@ -0,0 +1,141 @@
+]
+
+[tool.poetry.dependencies]
+python = ">=3.8.15, <3.12.0"  # don't change this
```
```suggestion
python = ">=3.10.0, <3.12.0"  # don't change this```

jafermarq reviewed (commented):
I few extra comments i forgot to add above

QinbinLi commented:
Hi @jafermarq ,

Thanks a lot! I have fixed all your comments. Please take another look. Thank you!

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,136 @@
+dataset: [CIFAR-10, CIFAR-100] # list of datasets you include in your baseline
+---
+
+# :warning:*_Title of your baseline_*
```
```suggestion
# Model-Contrastive Federated Learning
```

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,136 @@
+title: Model-Contrastive Federated Learning
+url: https://arxiv.org/abs/2103.16257
+labels: [data heterogeneity, image classification, cross-silo, constrastive-learning]
+dataset: [CIFAR-10, CIFAR-100] # list of datasets you include in your baseline
```
```suggestion
dataset: [CIFAR-10, CIFAR-100]
```

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,136 @@
+
+****Contributors:**** : Qinbin Li
+
+****Description:****: MOON requires to compute the model-contrastive loss in local training, which requires access to the local model of the previous round (Lines 14-17 of Algorithm 1 of the paper). Since `FlowerClient` does not preserve the states when starting a new round, we store the local models into the specified `model_dir` in local training indexed by the client ID, which will be loaded t ...
```
```suggestion
****Description:****: MOON requires to compute the model-contrastive loss in local training, which requires access to the local model of the previous round (Lines 14-17 of Algorithm 1 of the paper). Since currently `FlowerClient` does not preserve the states when starting a new round, we store the local models into the specified `model.dir` in local training indexed by the client ID, which will be loaded to the corresponding client in the next round.
```

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,136 @@
+
+## About this baseline
+
+****What’s implemented:**** : The code in this directory replicates the experiments in *Model-Contrastive Federated Learning* (Li et al., 2021), which proposed the MOON algorithm. Concretely ,it replicates the results of MOON for CIFAR-10 and CIFAR-100 in Table 1.
```
```suggestion
****What’s implemented:**** The code in this directory replicates the experiments in *Model-Contrastive Federated Learning* (Li et al., 2021), which proposed the MOON algorithm. Concretely ,it replicates the results of MOON for CIFAR-10 and CIFAR-100 in Table 1.
```

jafermarq replied:
Most of the paragraphs  in the `README.md` actually have this problem where the `:` is used in the part that's bold and also after. Please only use it in the **bold** part.

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,136 @@
+| MOON | 0.7107 | 0.6799 |
+| FedProx| 0.6852 | 0.6494 |
+
+You can find the curve comparing MOON and FedProx on CIFAR-10 below.
+
+![](_static/cifar10_moon_fedprox.png)
```
Could we here have a new sub-section titled `### Figure 6` that shows side-by-side the plot you already have and next to it the equivalent but for CIFAR-100? Also, please add a title and grid lines to the plots (for consistency with other baselines)

To make the plots side-by-side in your readme you can do something like:
```
<img alt="some meaningful description" src="_static/cifar10_moon_fedprox.png" width="400"/> <img alt="some meaningful description" src="_static/cifar10_moon_fedprox.png" width="400"/>
```

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,136 @@
+
+You can tune the hyperparameter `mu` for both MOON and FedProx by changing the configuration file in `conf`.
+
+You can also run the experiments in Figure 8 of the paper. To run MOON on CIFAR-100 with 50 clients (Figure 8(a) of the paper):
```
Similarly to my comment above, could put this for Figure 8 as a separate sub-section? Then below, again a side-by-side plot following the style of Figure 6.

QinbinLi replied:
Hi @jafermarq ,

Thank you! I have addressed other comments. For Figure 8, I'm rerunning the experiments and the experiment will take a long time. I will probably need 1-2 weeks to complete it. I'll upload the results once completed. You may go ahead with the current version if it's ok for you.

jafermarq commented on baselines/moon/pyproject.toml:
```diff
@@ -0,0 +1,142 @@
+matplotlib = "3.8.0"
+torch = { url = "https://download.pytorch.org/whl/cu116/torch-1.12.0%2Bcu116-cp310-cp310-linux_x86_64.whl"}
+torchvision = { url = "https://download.pytorch.org/whl/cu116/torchvision-0.13.0%2Bcu116-cp310-cp310-linux_x86_64.whl"}
+[tool.poetry.dev-dependencies]
```
```suggestion

[tool.poetry.dev-dependencies]
```

jafermarq reviewed (commented):
Hi @QinbinLi ,

Thanks for making the changes to your PR. I have listed a few minor ones. The main one is related to including the plots in the `README.md`. Ideallly, we'll have two 1x2 plots (one as in Figure 6 and one as in Figure 8 in your paper). 

You can remove the other plots that are currently in `_static` if they are not part of the `README.md`. 

Not a big review this time. Ping me when you are done with these. I'll re-run some of your experiments in the mean time and let you know if something is not quite right.

jafermarq commented on baselines/moon/pyproject.toml:
```diff
@@ -0,0 +1,143 @@
+version = "1.0.0"
+description = "Flower Baselines"
+license = "Apache-2.0"
+authors = ["The Flower Authors <hello@flower.dev>"]
```
You might want to add your contact details here. Also, please add the title of your paper to the `description` field above.

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,136 @@
+You can find the curve comparing MOON and FedProx on CIFAR-10 and CIFAR-100 below.
+
+
+<img alt="CIFAR-10" src="_static/cifar10_moon_fedprox.png" width="400"/> <img alt="CIFAR-100" src="_static/cifar100_moon_fedprox.png" width="400"/>
```
I'm thinking we can make these plots a bit larger. Could you try and see if with width=500 still they fit side by side?

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,136 @@
+
+# activate the environment
+poetry shell
+
```
```suggestion
```

jafermarq commented on baselines/moon/README.md:
```diff
@@ -0,0 +1,136 @@
+> Note: If you use this baseline in your work, please remember to cite the original authors of the paper as well as the Flower paper.
+
+
+****Paper:**** : [arxiv.org/abs/2103.16257](https://arxiv.org/abs/2103.16257)
```
There is a repeated `:` after the **bold** text. This issue happens other times below.

jafermarq reviewed (commented):
Hi @QinbinLi ,

Just a couple of minor comments.

Ping me once you get those pending results and i'll give a more in-depth (likely final) review to your baseline.

jafermarq commented:
Hi @QinbinLi, hope things are going well. Please let me know if there is anything you'd like discuss about your baseline. From our last discussion it seems everything is in place and only a few experiments need to be run and added to the `README.md`. Happy to help if some parts need a second pair of eyes or if you are struggling making the `./dev/test-baseline.sh` script pass.

QinbinLi commented:
Hi @jafermarq ,

I have updated the code according to your comments. I have added the results of Figure 8(a) ([https://arxiv.org/pdf/2103.16257.pdf](https://arxiv.org/pdf/2103.16257.pdf)) using MOON. The experiment using FedProx failed multiple times due to insufficient memory (others are using the same machine) and I launched another machine to run it. I need about 2 days to finish it.

Is it necessary to run Figure 8(b)? Figure 8(b) is more time-consuming than Figure 8(a).

jafermarq commented:
Hi @QinbinLi, it's fine not including Fig 8(b). Ping me once the `FedProx` is added to your Fig8(a). If possible, could you make the figures to use larger font size (just so it's more readable in the `README.md`. Many thanks

jafermarq commented on baselines/moon/_static/cifar100_moon_fedprox.png:
If possible, could we make the font larger for labels, ticks and legend? (so it's readable from the `README.md`)

jafermarq reviewed (commented):
Hi @QinbinLi ,

You'll notice i pushed some changes (I did just some formatting). We can start merging your baseline once you have added the `FedProx` curve to your Fig8a below in the readme. 

I think it would be great if you could add some comments to your `models.train_moon()` function (for instance state which line(s) doe each line in Algorithm 1 in your paper). This would be ideal for others trying to use (maybe even modify!) your MOON baseline. I left a comment below.

Also, I'm thinking is better to remove the log files in `_static`. They simply take too much space. If you are able to host these somewhere else, you are welcome to add a link to it in the `README.md` if you want.

Finally, I believe the `client_states` directory needs to be erased before running a new experiment, right? I see you remove it once the experiment is completed. But, what if the experiment crashes for some reason. We probably want to erase it before starting the simuilation again, right? 

Please ping me when you managed to add the final result and go through my comments.

QinbinLi commented:
Hi @jafermarq ,

I have addressed your comments and pushed the new results. Thanks!

jafermarq reviewed (approved):
@QinbinLi, great! thanks for the update. This is ready to be merged. I'll do this early in the week.
